nodecg-io-intellij/extension/intellij.ts

```diff
@@ -0,0 +1,978 @@
+        const result = await this.rawRequest("get_project", { project: projectName });
+        if (result != null) {
+            return new Project(this, result);
+        } else {
+            throw new Error("Project not found");
+        }
+    }
+
+    /**
+     * Gets the given URL as an IntelliJFile. If the file is not found, the promise is
+     * rejected.
+     */
+    async getFile(url: string): Promise<VirtualFile> {
+        const result = await this.rawRequest("vfs_get_by_url", { url: url });
+        if (result != null) {
+            return new VirtualFile(this, result);
+        } else {
+            throw new Error("File not found");
+        }
+    }
+
+    /**
+     * Runs the action with the given id. An ation is for example an button in the
+     * menu toolbar. A non-exhaustive list of actions can be found at
+     * https://centic9.github.io/IntelliJ-Action-IDs/
+     * An action place is a string that specifies from where the action call should
+     * be simulated. Available action places can be found at
+     * https://github.com/JetBrains/intellij-community/blob/master/platform/platform-api/src/com/intellij/openapi/actionSystem/ActionPlaces.java
+     * This also allows you to call action that are not enabled (grey and unclickable) at
+     * the moment. This could lead to unexpected behaviour. Use with caution
+     * Please note that this method executes an action from outside of a project. That
+     * means no project is available for the action. You might be looking for
+     * `IntelliJProject.action()` as most actions need a project.
+     * @param action The action id
+     * @param place The place from where the event is simulated.
+     * @param sync If this is set to true the methods blocks while the action is executing.
+     */
+    async action(action: string, place: string, sync: boolean): Promise<void> {
+        await this.rawRequest("run_action", {
+            action: action,
+            place: place,
+            sync: sync,
+            project: null,
+        });
+    }
+
+    /**
+     * Makes a raw request to IntelliJ
+     */
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any, @typescript-eslint/explicit-module-boundary-types
+    async rawRequest(method: string, data: any): Promise<any> {
+        const response = await fetch(this.address, {
+            method: "POST",
+            body: JSON.stringify({
+                method: method,
+                data: data,
+            }),
+        });
+        const json = await response.json();
+        if (json.success) {
+            return json.data;
+        } else if ("data" in json && "error_msg" in json.data) {
+            throw new Error(json.data.error_msg);
+        } else {
+            throw new Error("unknown");
+        }
+    }
+
+    equals(other: IntelliJ): boolean {
+        return this.address == other.address;
+    }
+}
+
+export class Project {
+    readonly intellij: IntelliJ;
```

Comment:
This should be made private as the user isn't expected to use it and it is just for internals. This also applies to all other classes that have a intellij instance.
```suggestion
    private readonly intellij: IntelliJ;
```

Reply:
I made the protected as there is a class that extends another (HistoryLabel) and there might be more of them in the future.